src/charm.py

```diff
@@ -46,25 +49,68 @@ def _on_install(self, _: ops.InstallEvent) -> None:
         init_config = state.BuilderInitConfig.from_charm(self)
         builder.initialize(init_config=init_config)
         self.unit.status = ops.ActiveStatus("Waiting for first image.")
-        self._run()
 
     @charm_utils.block_if_invalid_config(defer=False)
     def _on_config_changed(self, _: ops.ConfigChangedEvent) -> None:
         """Handle charm configuration change events."""
-        proxy.configure_aproxy(proxy=state.ProxyConfig.from_env())
         init_config = state.BuilderInitConfig.from_charm(self)
+        if not self._is_image_relation_ready_set_status(config=init_config.run_config):
+            return
+        proxy.configure_aproxy(proxy=state.ProxyConfig.from_env())
         builder.install_clouds_yaml(cloud_config=init_config.run_config.cloud_config)
         if builder.configure_cron(unit_name=self.unit.name, interval=init_config.interval):
             self._run()
         self.unit.status = ops.ActiveStatus()
 
     @charm_utils.block_if_invalid_config(defer=False)
-    def _on_run_action(self, _: ops.EventBase) -> None:
-        """Handle the run action event."""
+    def _on_image_relation_changed(self, _: ops.RelationChangedEvent) -> None:
+        """Handle charm image relation changed event."""
+        init_config = state.BuilderInitConfig.from_charm(self)
+        if not self._is_image_relation_ready_set_status(config=init_config.run_config):
+            return
+        proxy.configure_aproxy(proxy=state.ProxyConfig.from_env())
+        builder.install_clouds_yaml(cloud_config=init_config.run_config.cloud_config)
+        self._run()
+        self.unit.status = ops.ActiveStatus()
+
+    @charm_utils.block_if_invalid_config(defer=False)
+    def _on_run_action(self, event: ops.ActionEvent) -> None:
+        """Handle the run action event.
+
+        Args:
+            event: The run action event.
+        """
+        init_config = state.BuilderInitConfig.from_charm(self)
+        if not self._is_image_relation_ready_set_status(config=init_config.run_config):
+            event.fail("Image relation not yet ready.")
+            return
         self._run()
 
+    def _is_image_relation_ready_set_status(self, config: state.BuilderRunConfig) -> bool:
+        """Check if image relation is ready and set according status otherwise.
+
+        Args:
+            config: The image builder run configuration.
+
+        Returns:
+            Whether the image relation is ready.
+        """
+        if state.UPLOAD_CLOUD_NAME not in config.cloud_config["clouds"]:
+            self.unit.status = ops.BlockedStatus(f"{state.IMAGE_RELATION} integration required.")
+            return False
```

Comment:
This can lead to confusing BlockedStatus if there is already an integration with an older Github Runner Charm revision (where this data is not provided in the integration). Could you mention somewhere in the docs that a specific Github Runner revision is required (I think its rev 246: https://github.com/canonical/github-runner-operator/releases/tag/rev246)?

Reply:
This requires https://github.com/canonical/github-runner-operator/pull/362 to be merged first, so I think that would be > rev 249 - thanks for suggesting this, i'll push this with desired revision after the fix PR has been merged :)

Reply:
https://github.com/canonical/github-runner-operator/pull/362 has been merged

Reply:
Added release-notes, thank you!